src/operator/tensor/elemwise_binary_op.h

```diff
@@ -13,350 +13,669 @@
-  });
-}
-
-template<typename xpu, typename OP>
-void BinaryComputeWithHalf2(const nnvm::NodeAttrs& attrs,
-                            const OpContext& ctx,
-                            const std::vector<TBlob>& inputs,
-                            const std::vector<OpReqType>& req,
-                            const std::vector<TBlob>& outputs) {
-  MSHADOW_TYPE_SWITCH_WITH_HALF2(outputs[0].type_flag_, DType, {
-    BinaryCompute_<xpu, OP, DType>(attrs, ctx, inputs, req, outputs);
-  });
-}
-
-template<typename xpu, typename op>
-void BinaryLaunch(const nnvm::NodeAttrs& attrs,
-                  const OpContext& ctx,
-                  const std::vector<TBlob>& inputs,
-                  const std::vector<OpReqType>& req,
-                  const std::vector<TBlob>& outputs) {
-  using namespace mshadow;
-  using namespace mxnet_op;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-
-  CHECK_EQ(inputs.size(), 2U);
-  CHECK_EQ(outputs.size(), 1U);
-  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
-    Kernel<op, xpu>::Launch(s, outputs[0].Size(),
-      outputs[0].dptr<DType>(), inputs[0].dptr<DType>(), inputs[1].dptr<DType>());
-  });
-}
-
-template<typename OP, int Req >
-struct BinaryOpBackwardUseNone {
   template<typename DType>
-  MSHADOW_XINLINE static void Map(int i, DType* igrad, const DType* ograd) {
-    KERNEL_ASSIGN(igrad[i], Req, OP::Map(ograd[i]));
+  MSHADOW_XINLINE static void Map(int i, DType *out, const DType value) {
+    KERNEL_ASSIGN(out[i], Req, value);
   }
 };
 
-template<typename xpu, typename LOP, typename ROP, typename DType>
-void BinaryBackwardUseNone_(const nnvm::NodeAttrs& attrs,
-                            const OpContext& ctx,
-                            const std::vector<TBlob>& inputs,
-                            const std::vector<OpReqType>& req,
-                            const std::vector<TBlob>& outputs) {
-  using namespace mxnet_op;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-  int size = static_cast<int>((outputs[0].Size() + DataType<DType>::kLanes - 1)
-    /DataType<DType>::kLanes);
-  DType* lgrad_dptr = outputs[0].dptr<DType>();
-  DType* rgrad_dptr = outputs[1].dptr<DType>();
-  DType* ograd_dptr = inputs[0].dptr<DType>();
-  if (std::is_same<LOP, mshadow_op::identity>::value && req[0] == kWriteInplace) {
-    CHECK_EQ(ograd_dptr, lgrad_dptr);
-  } else if (req[0] != kNullOp) {
-    MXNET_ASSIGN_REQ_SWITCH(req[0], Req,
-      {Kernel<BinaryOpBackwardUseNone<LOP, Req>, xpu>::Launch(s, size, lgrad_dptr,
-        ograd_dptr);});
-  }
-  if (std::is_same<ROP, mshadow_op::identity>::value && req[1] == kWriteInplace) {
-    CHECK_EQ(ograd_dptr, rgrad_dptr);
-  } else if (req[1] != kNullOp) {
-    MXNET_ASSIGN_REQ_SWITCH(req[1], Req,
-      {Kernel<BinaryOpBackwardUseNone<ROP, Req>, xpu>::Launch(s, size, rgrad_dptr,
-        ograd_dptr);});
+/*! Gather binary operator functions into BinaryOp class */
+class BinaryOp : public OpBase
+{
```

Comment:
why change all function to static and wrap in a class? Would a namespace be sufficient?